MapleServer2/PacketHandlers/Game/BeautyHandler.cs

```diff
@@ -0,0 +1,611 @@
+                case BeautyShopType.Makeup:
+                    voucherTag = "beauty_makeup";
+                    break;
+                case BeautyShopType.Skin:
+                    voucherTag = "beauty_skin";
+                    break;
+                case BeautyShopType.Dye:
+                    voucherTag = "beauty_itemcolor";
+                    break;
+                default:
+                    session.Send(NoticePacket.Notice("Unknown Beauty Shop", NoticeType.FastText));
+                    return;
+            }
+
+            Item voucher = session.Player.Inventory.Items.FirstOrDefault(x => x.Value.Tag == voucherTag).Value;
+            if (voucher == null)
+            {
+                session.Send(NoticePacket.Notice(SystemNotice.ItemNotFound, NoticeType.FastText));
+                return;
+            }
+
+            session.Send(BeautyPacket.UseVoucher(voucher.Id, 1));
+            InventoryController.Consume(session, voucher.Uid, 1);
+        }
+
+        private static void PayWithShopTokenCost(GameSession session, BeautyMetadata beautyShop)
+        {
+            int cost;
+            if (beautyShop.SpecialCost != 0)
+            {
+                cost = beautyShop.SpecialCost;
+            }
+            else
+            {
+                cost = beautyShop.TokenCost;
+            }
+
+            switch (beautyShop.TokenType)
+            {
+                case ShopCurrencyType.Meso:
+                    session.Player.Wallet.Meso.Modify(-(cost));
+                    break;
+                case ShopCurrencyType.ValorToken:
+                    session.Player.Wallet.ValorToken.Modify(-(cost));
+                    break;
+                case ShopCurrencyType.Treva:
+                    session.Player.Wallet.Treva.Modify(-(cost));
+                    break;
+                case ShopCurrencyType.Rue:
+                    session.Player.Wallet.Rue.Modify(-(cost));
+                    break;
+                case ShopCurrencyType.HaviFruit:
+                    session.Player.Wallet.HaviFruit.Modify(-(cost));
+                    break;
+                case ShopCurrencyType.Meret:
+                case ShopCurrencyType.GameMeret:
+                    session.Player.Wallet.Meret.Modify(-(cost));
+                    break;
+                case ShopCurrencyType.EventMeret:
+                    session.Player.Wallet.RemoveMerets(cost);
+                    break;
+                case ShopCurrencyType.Item:
+                    Item itemCost = session.Player.Inventory.Items.FirstOrDefault(x => x.Value.Id == beautyShop.RequiredItemId).Value;
+                    if (itemCost.Amount < cost)
+                    {
+                        return;
+                    }
+                    InventoryController.Consume(session, itemCost.Uid, cost);
+                    break;
+                default:
+                    session.SendNotice($"Unknown currency: {beautyShop.TokenType}");
+                    break;
```

Comment:
Could add this to a function and reuse on `PayWithShopItemTokenCost`

Reply:
I'm not sure if I can. I'm using two different functions because the cost type and amount in `PayWithShopItemTokenCost` depends on that specific item's cost, hence why I'm passing through the item id instead of the shop.
But maybe I'm not seeing a way to do it cleaner.